### PR TITLE
[BYOM] Don't Crash on Invalid Payload

### PR DESCRIPTION
**Pull request by jonathansampson**
Addresses crashing behavior of https://github.com/brave/brave-browser/issues/43488

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

Configure a local server to respond to POST messages with the following:

```json
{"choices":[{"message":{"not_content":"value"}}]}
```
Configure that endpoint as a BYOM server, and attempt to use it.

### CODE REVIEW

**Comment by jonathansampson** [comment]
Just noticed an issue where the final response from the server could unintentionally trigger the "invalid response format" alert. I'll address that ASAP in another commit.

File: components/ai_chat/core/browser/engine/oai_api_client_unittest.cc [hunk 1]
```diff
@@ -195,4 +194,65 @@ TEST_F(OAIAPIUnitTest, PerformRequest) {
...
 }
 
+TEST_F(OAIAPIUnitTest, HandleInavlidResponseFormat) {
+  mojom::CustomModelOptionsPtr model_options = mojom::CustomModelOptions::New(
+      "test_api_key", 0, 0, 0, "test_system_prompt", GURL("https://test.com"),
+      "test_model");
+
+  // Simulate an invalid response format from the server
+  std::string invalid_server_response =
+      R"({"choices":[{"message":{"not_content":"value"}}]})";
```
Comment (diracdeltas) [inline components/ai_chat/core/browser/engine/oai_api_client_unittest.cc:204]: can we add a couple more invalid test cases like

`aaaaaaaaaaaaaaaaa`
`{choices: []}`
`{"choices": [{"message": {"content": []}]}`

Comment (diracdeltas) [reply components/ai_chat/core/browser/engine/oai_api_client_unittest.cc:204]: and also test empty server response body

File: components/ai_chat/core/browser/conversation_handler.cc [hunk 2]
```diff
@@ -1529,6 +1529,11 @@ void ConversationHandler::OnSuggestedQuestionsResponse(
         mojom::SuggestionGenerationStatus::HasGenerated;
     DVLOG(2) << "Got questions:" << base::JoinString(result.value(), "\n");
   } else {
+    // handle failure
+    if (result.error() != mojom::APIError::None) {
+      DVLOG(2) << __func__ << ": With error";
+      SetAPIError(std::move(result.error()));
+    }
```
Comment (yrliou) [inline components/ai_chat/core/browser/conversation_handler.cc:1536]: We need to add test coverage for this part in conversation_handler_unittest.cc.

File: components/ai_chat/core/browser/engine/oai_api_client.cc [hunk 3]
```diff
@@ -134,27 +129,30 @@ void OAIAPIClient::OnQueryCompleted(GenerationCompletedCallback callback,
...
-        completion = *message->FindString("content");
-      }
+    // We expect to find a list of completion choices in the response.
+    const base::Value::List* choices =
+        result.value_body().is_dict()
+            ? result.value_body().GetDict().FindList("choices")
+            : nullptr;
+
+    // Retrieve the first completion choice's content, if any.
+    const auto* message =
```
Comment (yrliou) [inline components/ai_chat/core/browser/engine/oai_api_client.cc:139]: content

File: components/ai_chat/core/browser/engine/oai_api_client_unittest.cc (hunk 1 above)
Comment (yrliou) [reply components/ai_chat/core/browser/engine/oai_api_client_unittest.cc:204]: In general, every check in the C++ code should have test case covered.

File: components/ai_chat/core/browser/engine/oai_api_client.cc [hunk 4]
```diff
@@ -134,27 +129,30 @@ void OAIAPIClient::OnQueryCompleted(GenerationCompletedCallback callback,
...
+      std::move(callback).Run(base::ok(std::move(*message)));
+      return;
     }
 
-    std::move(callback).Run(base::ok(std::move(completion)));
+    // If no completion was found, log and return an error.
+    // This situation may occur when the response format is unexpected.
+    DVLOG(2) << "No completion was found in response.";
+    std::move(callback).Run(
+        base::unexpected(mojom::APIError::InvalidResponseFormat));
```
Comment (yrliou) [inline components/ai_chat/core/browser/engine/oai_api_client.cc:155]: We should probably have a follow-up issue created to update other clients (conversation_api_client, remote_completion_client) to use this too.

**Comment by jonathansampson** [comment]
The scope of this PR has changed. Instead of addressing the larger task of informing the user of an invalid format in the server's response, this PR now addresses only the most crucial element of the associated issue: the crashing behavior. A follow-up issue (https://github.com/brave/brave-browser/issues/43536) will handle the UI aspects of informing the user when the server's payload deviated from the expected format.

File: components/ai_chat/core/browser/engine/oai_api_client.cc [hunk 5]
```diff
@@ -150,7 +150,11 @@ void OAIAPIClient::OnQueryCompleted(GenerationCompletedCallback callback,
           DVLOG(2) << "No message dict found in response.";
           return;
         }
-        completion = *message->FindString("content");
+
+        const std::string* content = message->FindString("content");
+        if (content) {
+          completion = *content;
+        }
```
Comment (yrliou) [inline components/ai_chat/core/browser/engine/oai_api_client.cc:157]: Please add something like `PerformRequest_BadResponse` or `PerformRequest_InvalidResponse` test case in oai_api_client_unittest and cover these checks.
